sqlglot/expressions.py

```diff
@@ -564,17 +579,20 @@ def transform(self, fun, *args, copy=True, **kwargs):
         Returns:
             The transformed tree.
         """
-        node = self.copy() if copy else self
-        new_node = fun(node, *args, **kwargs)
+        root = None
+        new_node = None
 
-        if new_node is None or not isinstance(new_node, Expression):
-            return new_node
-        if new_node is not node:
-            new_node.parent = node.parent
-            return new_node
+        for node in (self.copy() if copy else self).dfs(prune=lambda n: n is not new_node):
+            new_node = fun(node, *args, **kwargs)
 
-        replace_children(new_node, lambda child: child.transform(fun, *args, copy=False, **kwargs))
-        return new_node
+            if root:
+                if new_node is not node:
+                    node.replace(new_node)
+            else:
+                root = new_node
+
+        assert root
+        return root.assert_is(Expression)
```

Comment:
Curious why these `assert_is` calls were added, is it for mypy?

Reply:
yes